### DISCUSSION TITLE
System-level Build Dependencies: Does upgrading these break Linuxbrew compiled binaries?

### DISCUSSION THREAD

**Discussion by jwhite007** [Homebrew/discussions#2608] [Linux] [answered]
We use Linuxbrew on our systems at my workplace, mostly because we don't have root. Should I be worried about our IT upgrading system-level C libraries at some point?

It appears that Linuxbrew compiled binaries have many system-level C lib dependencies. At first, it appeared that these were all from the glibc library which may have made dealing with IT system upgrades more manageable as I could have asked them to ignore glibc upgrades only. I then noticed ```libcrypt.so.1 => /lib64/libcrypt.so.1``` in the output of ```ldd $(which perl)```. That perl was installed from the Linuxbrew bottle. It appears that it's going to be hard to predict which system-level resources will be used by Linuxbrew compiled binaries.

I seem to recall Linuxbrew being less dependent on system-level stuff. Has this been changing gradually over time?

```
$ brew config
HOMEBREW_VERSION: 3.3.6
ORIGIN: https://github.com/Homebrew/brew
HEAD: 32a3e1fe28f9372a6c6d5eca52129f92b0f00ba0
Last commit: 9 days ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: da3eb5d2b49e25f6a024fa7226de1cff17712715
Core tap last commit: 4 days ago
Core tap branch: master
HOMEBREW_PREFIX: /home/linuxbrew/.linuxbrew
HOMEBREW_CACHE: /home/linuxbrew/.cache/Homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_CORE_GIT_REMOTE: https://github.com/Homebrew/homebrew-core
HOMEBREW_EDITOR: /usr/bin/env vim
HOMEBREW_GITHUB_API_TOKEN: set
HOMEBREW_LOGS: /home/linuxbrew/.logs/Homebrew
HOMEBREW_MAKE_JOBS: 48
HOMEBREW_TEMP: /home/linuxbrew/.temp
Homebrew Ruby: 2.6.8 => /home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.8/bin/ruby
CPU: 48-core 64-bit broadwell
Clang: N/A
Git: 2.34.1 => /home/linuxbrew/.linuxbrew/bin/git
Curl: 7.61.1 => /bin/curl
Kernel: Linux 4.18.0-348.2.1.el8_5.x86_64 x86_64 GNU/Linux
OS: Red Hat Enterprise Linux release 8.5 (Ootpa) (Ootpa)
Host glibc: 2.28
/usr/bin/gcc: 8.5.0
/usr/bin/ruby: N/A
glibc: N/A
gcc@5: N/A
xorg: N/A
```

**Comment by Bo98** [comment, marked as answer]
It should just be glibc and system GCC (libstdc++ & libgcc_s). `libcrypt` (not to be confused with `libcrypto`) is a part of glibc.

But both have retained ABI compatibility for a while so updating glibc and GCC should not be a concern. We enforce a minimum version but no maximum for this reason.

If a situation arose where a future system glibc caused problems, then we could allow Homebrew's `glibc` to be installed. The binaries we ship in our bottles are already set up so that they will prioritise Homebrew's glibc if installed. People who have a system glibc that's too old already rely on this feature.

So you should be fine updating system libraries without it breaking anything. If anything does, then that's a bug.

**Reply by jwhite007** [reply]
@Bo98 

Thank you so much for your quick reply.

I should have done my due diligence on libcrypt. I did a cursory search initially and concluded that it wasn't part of glibc. Upon further inspection, I see that it is a part of glibc. Thanks for pointing that out.

I've read about glibc maintaining backwards compatibility so that upgrades don't break builds; however, I wasn't quite sure how this magic works. It appears to me that if one has a build pointing to, for instance ```libc.so.6 => /lib64/libc.so.6``` as does Linuxbrew's Python bottle, then an upgrade to /lib64/libc.so.7 would break Python's functionality with Python complaining that it can't find /lib64/libc.so.6. Am I missing something here?

I'm also glad that you mention being able to use Homebrew's glibc on older systems. Some of our systems, which we're in the process of decommissioning, are still running RHEL 6.6 which can only install up to glibc 2.12 using yum. I have glibc 2.23 installed via Linuxbrew. How do I convert to using this glibc? Is there an env variable that I can set, something like USE_HOMEBREW_GLIBC=1 or such?

Right now Linuxbrew is pretty dysfunctional on our older systems.
```
$ brew list
Error: Vendored tools require system Glibc 2.13 or later (yours is 2.12).
Error: Failed to upgrade Homebrew Portable Ruby!
If there's no Homebrew Portable Ruby available for your processor:
- install Ruby 2.6.8 with your system package manager (or rbenv/ruby-build)
- make it first in your PATH
- try again
```
Cheers

**Reply by Bo98** [reply]
> then an upgrade to /lib64/libc.so.7 would break Python's functionality with Python complaining that it can't find /lib64/libc.so.6. Am I missing something here?

This is correct, but it has been at ABI version 6 for nearly 25 years now so you're unlikely to see a version 7 anytime soon. It would be a major update in the Linux community if that were to happen.

> How do I convert to using this glibc? Is there an env variable that I can set, something like USE_HOMEBREW_GLIBC=1 or such?

In terms of individual formulae, it should work automatically.

However when it comes to bootstrapping Homebrew itself, it requires a Ruby 2.6 installation and we only build one using glibc 2.13 (and then once that's running, we automatically install 2.23 where needed). It's a bit of a chicken and egg scenario. While we don't really support system glibc <2.13, you might get it working if you find another means of installing Ruby 2.6.

**Reply by jwhite007** [reply]
@Bo98,

You gave me hope that I might be able to resuscitate Linuxbrew on our older systems which have glibc 2.12. I was hoping to bootstrap Linuxbrew as you suggested by just installing a functional Ruby 2.6, followed by installing glibc with Linuxbrew and using that for future builds and Linuxbrew bottles; however, it doesn't seem to work for me.
```
$ /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
Homebrew requires Ruby 2.6 which was not found on your system.
Homebrew portable Ruby requires Glibc version 2.13 or newer,
and your Glibc version is too old. See:
  https://docs.brew.sh/Homebrew-on-Linux#requirements
Please install Ruby 2.6 and add its location to your PATH.

$ ruby --version
ruby 2.6.1p33 (2019-01-30 revision 66950) [x86_64-linux]
```

Any suggestions would be much appreciated. Thanks.

**Reply by jwhite007** [reply]
@Bo98 

Ok. The problem may be that it's not specifically Ruby 2.6.8. There seems to be some inconsistency in how Linuxbrew warns about this. The installer simply states to install and put Ruby 2.6 in one's path; however, I just noticed in the ```brew list``` output above from the dysfunctional Linuxbrew that it states that Ruby 2.6.8 is needed. Unfortunately Ruby 2.6.8 won't build on a system with glibc 2.12 either.

**Reply by Bo98** [reply]
Any 2.6 should work. Here's the test it does: https://github.com/Homebrew/install/blob/2252546b0bab8d2557992f4b2c146f6c0e4ec2fe/install.sh#L304-L306
